Userland/Applications/PixelPaint/HistogramWidget.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2022, Torsten Engelmann
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "HistogramWidget.h"
+#include "Image.h"
+#include "ImageEditor.h"
+#include "Layer.h"
+#include <LibGUI/Painter.h>
+#include <LibGfx/Palette.h>
+#include <LibGfx/Path.h>
+
+REGISTER_WIDGET(PixelPaint, HistogramWidget);
+
+namespace PixelPaint {
+
+HistogramWidget::HistogramWidget()
+{
+    set_height(65);
+}
+
+HistogramWidget::~HistogramWidget()
+{
+    if (m_image)
+        m_image->remove_client(*this);
+}
+
+void HistogramWidget::set_image(Image* image)
+{
+    if (m_image == image)
+        return;
+    if (m_image)
+        m_image->remove_client(*this);
+    m_image = image;
+    if (m_image)
+        m_image->add_client(*this);
+
+    rebuild_histogram_data();
+}
+
+void HistogramWidget::rebuild_histogram_data()
+{
+    if (m_image) {
+        auto bitmap_or_error = m_image->try_compose_bitmap(Gfx::BitmapFormat::BGRA8888);
+        if (bitmap_or_error.is_error())
+            return;
+        auto full_bitmap = bitmap_or_error.release_value();
+
+        m_data.Red.clear();
+        m_data.Green.clear();
+        m_data.Blue.clear();
```

Comment:
It's probably much faster to use `clear_with_capacity()` than freeing the memory and allocating in a loop below.